libp2p/Host.cpp

```diff
@@ -749,46 +733,54 @@ void Host::startedWorking()
     // initialization (e.g. start capability threads, start TCP listener, and kick off timers)
     asserts(!m_timer);
 
-    {
-        // prevent m_run from being set to true at same time as set to false by stop()
-        // don't release mutex until m_timer is set so in case stop() is called at same
-        // time, stop will wait on m_timer and graceful network shutdown.
-        Guard l(x_runTimer);
-        // create deadline timer
-        m_timer.reset(new io::deadline_timer(m_ioService));
-        m_run = true;
-    }
+    // create deadline timer
+    m_timer.reset(new io::deadline_timer(m_ioService));
+    m_run = true;
 
-    // start capability threads (ready for incoming connections)
-    for (auto const& h: m_capabilities)
-        h.second->onStarting();
-    
-    // try to open acceptor (todo: ipv6)
-    int port = Network::tcp4Listen(m_tcp4Acceptor, m_netConfig);
-    if (port > 0)
+    if (m_capabilities.size())
     {
-        m_listenPort = port;
-        determinePublic();
-        runAcceptor();
+        // start capability threads (ready for incoming connections)
+        for (auto const& h : m_capabilities)
+            h.second->onStarting();
+
+        // try to open acceptor (todo: ipv6)
+        int port = Network::tcp4Listen(m_tcp4Acceptor, m_netConfig);
+        if (port > 0)
+        {
+            // Port may have been changed if the specified port was already in use
+            m_netConfig.listenPort = m_listenPort = port;
+            determinePublic();
+            runAcceptor();
+        }
+        else
+            LOG(m_logger) << "p2p.start.notice id: " << id()
+                          << " TCP Listen port is invalid or unavailable.";
     }
-    else
-        LOG(m_logger) << "p2p.start.notice id: " << id() << " TCP Listen port is invalid or unavailable.";
-
-    auto nodeTable = make_shared<NodeTable>(
-        m_ioService,
-        m_alias,
-        NodeIPEndpoint(bi::address::from_string(listenAddress()), listenPort(), listenPort()),
-        m_netConfig.discovery,
-        m_netConfig.allowLocalDiscovery
-    );
-    nodeTable->setEventHandler(new HostNodeTableHandler(*this));
+
+    auto nodeTable = make_shared<NodeTable>(m_ioService, m_alias,
+        // Use data from network configuration rather than Host's settings because p2p might not
```

Comment:
I'm a little confused about the purpose of `determinePublic()` and `m_tcpPublic`:
* `determinePublic()` configures `m_tcpPublic` but `m_tcpPublic `doesn't appear to actually be used for anything i.e. it's not tied to the host's tcp acceptor - in `Network::tcp4Listen` (which is where we create the tcp acceptor) we create a new endpoint and use it when opening the tcp acceptor, we don't use `m_tcpPublic` here: https://github.com/ethereum/aleth/blob/91787f239566501a2b23f17946618ddf6a3cee15/libp2p/Network.cpp#L141-L156

* `listenAddress()` and `listenPort()` are host-specific i.e. they expose the host's listen information - I've updated them to check if capabilities exist and if so, they return the host's listen information but if not, they call into new NodeTable functions to obtain the node table's listen information. 
*` listenAddress() `and `listenPort()` query the host's listen information from the `m_tcpPublic `endpoint - since `m_tcpPublic `isn't actually used for anything, they should instead presumably query the host's tcp acceptor for listen information (e.g. they can obtain the acceptor's local endpoint via `boost::asio::ip::tcp::acceptor::local_endpoint()` and query it for the listen information)?

* I'm not completely clear on the difference between a public vs private IP address from Aleth's perspective - we enable the user to supply both (via the `--listen-ip` and `--public-ip` args) and both are saved in the network config (`NetworkConfig::listenIPAddress` and `NetworkConfig::publicIPAddress`). Why do we need to let the user supply an IP address at all, is this in case the user's machine has more than 1 network interface?

Reply:
`m_tcpPublic` is used by the NodeTable as our IP we advertise to others (`NodeIPEndpoint` passed to `NodeTable` constructor uses `listenAddress()` which uses `m_tcpPublic`)

Reply:
Ah of course, then presumably if we don’t advertise a public IP we can only connect to others but no one else can connect to us? Or does NAT magically take care of that?

Reply:
>  However, why do we need to support both options rather than just a single IP

`listenIPAddress` is used to listen to new connections (see `Network::tcp4Listen` function), it's our address as we see it behind the NAT
`publicIPAddress` is what we give out to others through Discovery, it's our address as seen out of NAT

Reply:
> Ah of course, then presumably if we don’t advertise a public IP we can only connect to others but no one else can connect to us? 

More exactly they can't discover our public IP automatically; they still can add us as a peer manually if they know our public address through some other way.